providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,746 @@
+
+            /* Only include this param if it's got an associated value */
+            if (data_size > 0) {
+                if (plen[j] < data_size)
+                    goto err;
+                lprm[k] = params[j];
+                lprm[k].data = pptr[j];
+                lprm[k++].data_size = data_size;
+                pptr[j] += data_size;
+                plen[j] -= data_size;
+            }
+        }
+        /* Don't bother if there are no params */
+        if (k > 0) {
+            lprm[k] = OSSL_PARAM_construct_end();
+            ctx = EVP_PKEY_CTX_new_from_name(key->libctx,
+                                             key->info->alg[i].name,
+                                             key->propq);
+            if (ctx == NULL
+                    || EVP_PKEY_fromdata_init(ctx) <= 0
+                    || EVP_PKEY_fromdata(ctx, key->keys + i, selection,
+                                         lprm) <= 0)
+                goto err;
+        }
+    }
+    res = 1;
+ err:
+    EVP_PKEY_CTX_free(ctx);
+    OPENSSL_free(lprm);
+    OPENSSL_free(pptr);
+    OPENSSL_free(plen);
+    return res;
+}
+
+typedef struct {
+    unsigned int alg_n;
+    unsigned int param_n;
+    OSSL_PARAM params[MAX_HYBRID_PARAMS + 1];
+    struct {
+        unsigned char *data;
+        size_t data_size;
+    } *d[MAX_HYBRID_PARAMS][MAX_HYBRID_ALGS];
+} HYBRID_EXPORT_CB_ARG;
+
+static int hybrid_export_cb(const OSSL_PARAM params[], void *varg)
+{
+    HYBRID_EXPORT_CB_ARG *arg = (HYBRID_EXPORT_CB_ARG *)varg;
+    unsigned char *data;
+    unsigned int i;
+    int pos;
+    OSSL_PARAM *p;
+
+    /* Record information about each OSSL_PARAM this algorithm exports */
+    for (i = 0; params[i].key != NULL; i++) {
+        if ((p = OSSL_PARAM_locate(arg->params, params[i].key)) == NULL) {
+            if (arg->param_n == MAX_HYBRID_PARAMS)
+                return 0;
+            p = arg->params + arg->param_n++;
+            memcpy(p, params + i, sizeof(*p));
+            p->data = NULL;
+            p->data_size = 0;
+        }
+        data = OPENSSL_memdup(params->data, params->data_size);
+        if (data == NULL)
+            return 0;
+        pos = p - arg->params;
+        arg->d[pos][arg->alg_n]->data = data;
+        arg->d[pos][arg->alg_n]->data_size = params->data_size;
+    }
+    arg->alg_n++;
+    return 1;
+}
+
+int ossl_hybrid_kmgmt_export(void *keydata, int selection,
+                             OSSL_CALLBACK *param_cb, void *cbarg)
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)keydata;
+    HYBRID_EXPORT_CB_ARG arg;
+    unsigned int i, j;
+    int ret = 0;
+
+    if (!ossl_prov_is_running() || key == NULL)
+        return 0;
+
+    /* Gather all the key data */
+    memset(&arg, 0, sizeof(arg));
+    for (i = 0; i < key->info->num_algs; i++)
+        if (!EVP_PKEY_export(key->keys[i], selection,
+                             &hybrid_export_cb, &arg))
+            goto err;
+
+    /* Package each OSSL_PARAM */
```

Comment:
So export/import uses different format for the priv/pub params than get_params/set_params. That is really horrible. What about not supporting priv and pub in get/set_params at all? And just support the encoded public key param there. That should be sufficient for TLS, shouldn't it?

Reply:
Good suggestion.  TLS only uses OSSL_PARAM_PUB_KEY.  This has to be in the format specified by the drafts (raw concatenation of public keys).  The import/export has multiple params, so isn't really the same.

Reply:
That's a good suggestion and I've implemented it.
The format different is reasonable given that different param names are being used.